File: url-shortener/src/domain/id_generator.rs
//! ID generation using Base62 encoding

use rand::Rng;
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Arc;

/// Base62 character set: 0-9, a-z, A-Z
const CHARSET: &[u8] = b"0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
const BASE: u64 = 62;

/// ID Generator for creating short codes
#[derive(Debug, Clone)]
pub struct IdGenerator {
    /// Current counter value
    counter: Arc<AtomicU64>,

    /// Code length
    code_length: usize,

    /// Range start (for distributed counter allocation)
    range_start: u64,

    /// Range end
    range_end: u64,
}

impl IdGenerator {
    /// Create a new ID generator
    pub fn new(code_length: usize) -> Self {
        Self {
            counter: Arc::new(AtomicU64::new(0)),
            code_length,
            range_start: 0,
            range_end: u64::MAX,
        }
    }

    /// Create a new ID generator with a specific range
    /// Used for distributed systems where each instance gets a range
    pub fn with_range(code_length: usize, range_start: u64, range_end: u64) -> Self {
        Self {
            counter: Arc::new(AtomicU64::new(range_start)),
            code_length,
            range_start,
            range_end,
        }
    }

    /// Generate a new short code
    pub fn generate(&self) -> String {
        // Get next counter value
        let counter = self.counter.fetch_add(1, Ordering::SeqCst);

        // Check if we've exhausted our range
        if counter >= self.range_end {
            // In production, this would trigger a range refresh from DynamoDB
            // For now, we'll wrap around with a random offset
            let random_offset: u64 = rand::thread_rng().gen();
            self.counter.store(self.range_start + (random_offset % 1000), Ordering::SeqCst);
        }

        // Encode to base62
        self.encode(counter)
    }

    /// Generate a random short code (fallback for collisions)
    pub fn generate_random(&self) -> String {
        let mut rng = rand::thread_rng();
        let mut code = String::with_capacity(self.code_length);

        for _ in 0..self.code_length {
            let idx = rng.gen_range(0..BASE as usize);
            code.push(CHARSET[idx] as char);
        }

        code
    }

    /// Encode a number to base62 string
    pub fn encode(&self, mut num: u64) -> String {
        if num == 0 {
            return self.pad("0".to_string());
        }

        let mut result = String::new();

        while num > 0 {
            let remainder = (num % BASE) as usize;
            result.insert(0, CHARSET[remainder] as char);
            num /= BASE;
        }

        self.pad(result)
    }

    /// Decode a base62 string to number
    pub fn decode(&self, code: &str) -> Option<u64> {
        let mut result: u64 = 0;

        for c in code.chars() {
            let value = match c {
                '0'..='9' => (c as u64) - ('0' as u64),
                'a'..='z' => (c as u64) - ('a' as u64) + 10,
                'A'..='Z' => (c as u64) - ('A' as u64) + 36,
                _ => return None,
            };

            result = result.checked_mul(BASE)?.checked_add(value)?;
        }

        Some(result)
    }

    /// Pad the code to the required length
    fn pad(&self, mut code: String) -> String {
        while code.len() < self.code_length {
            code.insert(0, '0');
        }
        code
    }

    /// Validate a short code format
    pub fn is_valid_code(&self, code: &str) -> bool {
        if code.len() != self.code_length {
            return false;
        }

        code.chars().all(|c| matches!(c, '0'..='9' | 'a'..='z' | 'A'..='Z'))
    }

    /// Validate a custom alias
    pub fn is_valid_custom_alias(alias: &str, max_length: usize) -> bool {
        if alias.len() < 4 || alias.len() > max_length {
            return false;
        }

        // Must be alphanumeric or hyphens
        if !alias.chars().all(|c| c.is_alphanumeric() || c == '-') {
            return false;
        }

        // Cannot start or end with hyphen
        if alias.starts_with('-') || alias.ends_with('-') {
            return false;
        }

        // Cannot have consecutive hyphens
        if alias.contains("--") {
            return false;
        }

        true
    }
}

impl Default for IdGenerator {
    fn default() -> Self {
        Self::new(7) // Default to 7-character codes
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_encode_decode() {
        let gen = IdGenerator::new(7);

        // Test various numbers
        let test_cases = [0u64, 1, 61, 62, 100, 1000, 1_000_000, 1_234_567_890];

        for num in test_cases {
            let encoded = gen.encode(num);
            let decoded = gen.decode(&encoded).unwrap();
            assert_eq!(decoded, num, "Failed for {}: encoded as {}", num, encoded);
        }
    }

    #[test]
    fn test_code_length() {
        let gen = IdGenerator::new(7);

        for _ in 0..100 {
            let code = gen.generate();
            assert_eq!(code.len(), 7);
        }
    }

    #[test]
    fn test_uniqueness() {
        let gen = IdGenerator::new(7);
        let mut codes = std::collections::HashSet::new();

        for _ in 0..10000 {
            let code = gen.generate();
            assert!(!codes.contains(&code), "Duplicate code generated: {}", code);
            codes.insert(code);
        }
    }

    #[test]
    fn test_valid_custom_alias() {
        assert!(IdGenerator::is_valid_custom_alias("my-link", 20));
        assert!(IdGenerator::is_valid_custom_alias("mylink123", 20));
        assert!(IdGenerator::is_valid_custom_alias("test", 20));

        assert!(!IdGenerator::is_valid_custom_alias("ab", 20)); // Too short
        assert!(!IdGenerator::is_valid_custom_alias("-mylink", 20)); // Starts with hyphen
        assert!(!IdGenerator::is_valid_custom_alias("mylink-", 20)); // Ends with hyphen
        assert!(!IdGenerator::is_valid_custom_alias("my--link", 20)); // Consecutive hyphens
        assert!(!IdGenerator::is_valid_custom_alias("my_link", 20)); // Underscore not allowed
    }
}

File: url-shortener-java/src/main/java/com/urlshortener/service/IdGenerator.java
package com.urlshortener.service;

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import jakarta.annotation.PostConstruct;
import java.security.SecureRandom;
import java.util.concurrent.atomic.AtomicLong;

/**
 * ID Generator using Base62 encoding with distributed range allocation.
 *
 * This implements the same algorithm as the Rust version:
 * - Each instance gets a range of IDs (default 1M)
 * - Local atomic counter for zero-coordination writes
 * - Range refresh from DynamoDB when threshold reached
 * - Base62 encoding: 0-9, a-z, A-Z (62 chars)
 *
 * Capacity: 62^7 = 3,521,614,606,208 unique codes
 * At 500M/month = 7,000+ years of capacity
 */
@Slf4j
@Component
public class IdGenerator {

    private static final String CHARSET = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
    private static final int BASE = 62;

    @Value("${url-shortener.code-length:7}")
    private int codeLength;

    @Value("${url-shortener.id-generator.range-size:1000000}")
    private long rangeSize;

    @Value("${url-shortener.id-generator.prefetch-threshold:0.9}")
    private double prefetchThreshold;

    private final AtomicLong counter = new AtomicLong(0);
    private volatile long rangeStart = 0;
    private volatile long rangeEnd = Long.MAX_VALUE;
    private final SecureRandom random = new SecureRandom();

    // For distributed mode - would be injected
    private CounterService counterService;

    @PostConstruct
    public void init() {
        log.info("IdGenerator initialized with code length: {}, range size: {}",
                 codeLength, rangeSize);
    }

    /**
     * Set counter service for distributed range allocation
     */
    public void setCounterService(CounterService counterService) {
        this.counterService = counterService;
    }

    /**
     * Initialize with a specific range (for distributed deployment)
     */
    public void initializeRange(long start, long end) {
        this.rangeStart = start;
        this.rangeEnd = end;
        this.counter.set(start);
        log.info("Range initialized: {} to {}", start, end);
    }

    /**
     * Generate a new unique short code
     */
    public String generate() {
        long value = counter.getAndIncrement();

        // Check if we need to refresh the range
        if (value >= rangeEnd) {
            synchronized (this) {
                // Double-check after acquiring lock
                if (counter.get() >= rangeEnd) {
                    refreshRange();
                    value = counter.getAndIncrement();
                }
            }
        }

        // Check if approaching threshold (async prefetch)
        if (shouldPrefetch(value)) {
            asyncPrefetchRange();
        }

        return encode(value);
    }

    /**
     * Generate a random short code (fallback for collisions)
     */
    public String generateRandom() {
        StringBuilder sb = new StringBuilder(codeLength);
        for (int i = 0; i < codeLength; i++) {
            sb.append(CHARSET.charAt(random.nextInt(BASE)));
        }
        return sb.toString();
    }

    /**
     * Encode a number to Base62 string
     */
    public String encode(long num) {
        if (num == 0) {
            return pad("0");
        }

        StringBuilder sb = new StringBuilder();
        while (num > 0) {
            int remainder = (int) (num % BASE);
            sb.insert(0, CHARSET.charAt(remainder));
            num /= BASE;
        }

        return pad(sb.toString());
    }

    /**
     * Decode a Base62 string to number
     */
    public Long decode(String code) {
        if (code == null || code.isEmpty()) {
            return null;
        }

        long result = 0;
        for (char c : code.toCharArray()) {
            int value = charToValue(c);
            if (value < 0) {
                return null;
            }
            result = result * BASE + value;
        }
        return result;
    }

    /**
     * Validate a short code format
     */
    public boolean isValidCode(String code) {
        if (code == null || code.length() != codeLength) {
            return false;
        }
        return code.chars().allMatch(c ->
            (c >= '0' && c <= '9') ||
            (c >= 'a' && c <= 'z') ||
            (c >= 'A' && c <= 'Z')
        );
    }

    /**
     * Validate a custom alias
     */
    public boolean isValidCustomAlias(String alias, int maxLength) {
        if (alias == null || alias.length() < 4 || alias.length() > maxLength) {
            return false;
        }
        // Must be alphanumeric or hyphens
        if (!alias.matches("^[a-zA-Z0-9-]+$")) {
            return false;
        }
        // Cannot start or end with hyphen
        if (alias.startsWith("-") || alias.endsWith("-")) {
            return false;
        }
        // Cannot have consecutive hyphens
        if (alias.contains("--")) {
            return false;
        }
        return true;
    }

    /**
     * Get current counter value (for monitoring)
     */
    public long getCurrentCounter() {
        return counter.get();
    }

    /**
     * Get remaining IDs in current range
     */
    public long getRemainingInRange() {
        return Math.max(0, rangeEnd - counter.get());
    }

    // Private helper methods

    private String pad(String code) {
        if (code.length() >= codeLength) {
            return code;
        }
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < codeLength - code.length(); i++) {
            sb.append('0');
        }
        sb.append(code);
        return sb.toString();
    }

    private int charToValue(char c) {
        if (c >= '0' && c <= '9') {
            return c - '0';
        } else if (c >= 'a' && c <= 'z') {
            return c - 'a' + 10;
        } else if (c >= 'A' && c <= 'Z') {
            return c - 'A' + 36;
        }
        return -1;
    }

    private boolean shouldPrefetch(long value) {
        long threshold = rangeStart + (long) ((rangeEnd - rangeStart) * prefetchThreshold);
        return value >= threshold && counterService != null;
    }

    private void asyncPrefetchRange() {
        // In production, this would be async
        log.debug("Would prefetch next range asynchronously");
    }

    private void refreshRange() {
        if (counterService != null) {
            // Production: Get new range from DynamoDB
            try {
                long[] newRange = counterService.allocateRange(rangeSize);
                this.rangeStart = newRange[0];
                this.rangeEnd = newRange[1];
                this.counter.set(rangeStart);
                log.info("Range refreshed from counter service: {} to {}", rangeStart, rangeEnd);
            } catch (Exception e) {
                log.error("Failed to refresh range, using random fallback", e);
                // Fallback: use random offset
                fallbackRangeRefresh();
            }
        } else {
            // Local mode: just wrap around with random offset
            fallbackRangeRefresh();
        }
    }

    private void fallbackRangeRefresh() {
        long randomOffset = Math.abs(random.nextLong()) % rangeSize;
        this.rangeStart = randomOffset;
        this.rangeEnd = rangeStart + rangeSize;
        this.counter.set(rangeStart);
        log.warn("Using fallback range: {} to {}", rangeStart, rangeEnd);
    }

    /**
     * Interface for distributed counter service
     */
    public interface CounterService {
        /**
         * Atomically allocate a range of IDs
         * @param size Number of IDs to allocate
         * @return Array of [start, end) for the allocated range
         */
        long[] allocateRange(long size);
    }
}
